lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php

```diff
@@ -114,31 +114,127 @@ public function __construct($dbh, $statement, OCI8Connection $conn)
      * @param string $statement The SQL statement to convert.
      *
      * @return string
+     * @throws \Doctrine\DBAL\Driver\OCI8\OCI8Exception
      */
     static public function convertPositionalToNamedPlaceholders($statement)
     {
-        $count = 1;
-        $inLiteral = false; // a valid query never starts with quotes
-        $stmtLen = strlen($statement);
-        $paramMap = array();
-        for ($i = 0; $i < $stmtLen; $i++) {
-            if ($statement[$i] == '?' && !$inLiteral) {
-                // real positional parameter detected
-                $paramMap[$count] = ":param$count";
-                $len = strlen($paramMap[$count]);
-                $statement = substr_replace($statement, ":param$count", $i, 1);
-                $i += $len-1; // jump ahead
-                $stmtLen = strlen($statement); // adjust statement length
-                ++$count;
-            } elseif ($statement[$i] == "'" || $statement[$i] == '"') {
-                $inLiteral = ! $inLiteral; // switch state!
+        $fragmentOffset = $tokenOffset = 0;
+        $fragments = $paramMap = array();
+        $currentLiteralDelimiter = null;
+
+        do {
+            if (!$currentLiteralDelimiter) {
+                $result = self::findPlaceholderOrOpeningQuote(
+                    $statement,
+                    $tokenOffset,
+                    $fragmentOffset,
+                    $fragments,
+                    $currentLiteralDelimiter,
+                    $paramMap
+                );
+            } else {
+                $result = self::findClosingQuote($statement, $tokenOffset, $currentLiteralDelimiter);
             }
+        } while ($result);
+
+        if ($currentLiteralDelimiter) {
+            throw new OCI8Exception(sprintf(
+                'The statement contains non-terminated string literal starting at offset %d',
+                $tokenOffset - 1
+            ));
         }
 
+        $fragments[] = substr($statement, $fragmentOffset);
+        $statement = implode('', $fragments);
+
         return array($statement, $paramMap);
     }
 
     /**
+     * Finds next placeholder or opening quote.
+     *
+     * @param string $statement The SQL statement to parse
+     * @param string $tokenOffset The offset to start searching from
+     * @param int $fragmentOffset The offset to build the next fragment from
+     * @param array $fragments Resulting query fragments
+     * @param string|null $currentLiteralDelimiter Current literal delimiter
+     * @param array $paramMap Parameter map
+     * @return bool Whether the token was found
+     */
+    private static function findPlaceholderOrOpeningQuote(
+        $statement,
+        &$tokenOffset,
```

Comment:
I think we can move all the parsing logic (not only the by-ref variables) into a different class. It might look like:
```php
$foo = new Foo($sql);
$newSql = $foo->getSql();
$paramMap = $foo->getParameterMap();
```

But I don't like that it only does `get*()` but doesn't do `do*()`. Maybe it could be a proxy (or sub-) class which would translate all messages with positional parameters (`__construct()`, `bind*()`) to the other object which only understands the named ones?